Classes/RealUrl/TitleMapping.php

```diff
@@ -0,0 +1,233 @@
+                break;
+            default:
+                $title = '';
+                break;
+        }
+        // make it url friendly
+        return $this->utility->convertToSafeString($title ?: (string)$id);
+    }
+
+    /**
+     * Decode title to id
+     *
+     * @param string $title
+     * @return int id or -1 if no id was found!
+     */
+    protected function decodeTitleToId(string $title): int
+    {
+        $controller = $this->getControllerAliasFromUrlDecoder();
+        $repository = $this->getRepositoryForController($controller);
+        switch ($controller) {
+            case 'lebenslagen':
+                /** @var LebenslagenRepository $repository */
+                $id = $this->findIdRecursively($title, 'name', $repository->getAll());
+                break;
+            case 'leistungen':
+                /** @var LeistungenRepository $repository */
+                $id = $this->findIdRecursively($title, 'displayName', $repository->getAll());
+                break;
+            case 'organisationseinheiten':
+                /** @var OrganisationseinheitenRepository $repository */
+                $id = $this->findIdRecursively($title, 'name', $repository->getAll());
+                break;
+            default:
+                $id = -1;
+                break;
+        }
+        return $id;
+    }
+
+    /**
+     * Find id of a Service BW item recursively
+     *
+     * @param string $search name of the item e.g. Stadtverwaltung
+     * @param string $key where the name is stored
+     * @param array $items e.g. from $repository->getAll() may includes _children items
+     * @return int the id or -1 if no id was found!
+     */
+    protected function findIdRecursively(string $search, string $key, array $items): int
+    {
+        foreach ($items as $item) {
+            if (array_key_exists($key, $item) && $this->utility->convertToSafeString($item[$key]) === $search) {
+                return (int)$item['id'];
+            }
+            // loop recursively if current item has children
+            if (array_key_exists('_children', $item)) {
+                $id = $this->findIdRecursively($search, $key, $item['_children']);
+                if ($id !== -1) {
+                    return $id;
+                }
+            }
+        }
+        return -1;
+    }
+
+    /**
+     * Extracts the controller alias from url decoders original path
+     *
+     * @return string
+     * @throws InvalidPathException
+     * @throws \ReflectionException
+     */
+    protected function getControllerAliasFromUrlDecoder(): string
+    {
+        $urlDecoderReflection = new \ReflectionClass($this->urlDecoder);
+        $reflectionProperty = $urlDecoderReflection->getProperty('originalPath');
+        $reflectionProperty->setAccessible(true);
+        $originalPath = $reflectionProperty->getValue($this->urlDecoder);
+        preg_match('/\/service\-bw\/([a-zA-Z0-9\-]+)\//', $originalPath, $matches);
```

Comment:
If you use `@` you don't need to escape all `/`

Reply:
Use alnum instead: http://php.net/manual/de/regexp.reference.character-classes.php

Reply:
http://php.net/manual/de/regexp.reference.subpatterns.php
Quote: It is possible to name a subpattern using the syntax (?P<name>pattern). This subpattern will then be indexed in the matches array by its normal numeric position and also by name

Makes your live easier